### ISSUE TITLE
Grading assignments is creating duplicate announcements for students

### ISSUE THREAD

**Issue by maskib** [UM-USElab/gradecraft-development#4252]
### Expected Behavior

Each grade submitted by an instructor should result in 1 announcement created on a student's account. 

### Actual Behavior

Currently this is triggering several announcements for each grade. The number seems to vary, but they continue to appear, within the few minutes after the grade was originally submitted. 

### Steps to reproduce

1. Create an assignment
2. Grade a student on that assignment
3. View the student's account and check the number of announcements made related to that one grade event. 

<img width="1552" alt="Screen Shot 2019-05-13 at 4 52 40 PM" src="https://user-images.githubusercontent.com/38505166/57653606-3b342100-75a0-11e9-8c3c-2c03297c8c91.png">

**Comment by chcholman** [comment]
➤ Michael Skib commented:

@card Not happening outside of staging

**Comment by chcholman** [comment]
➤ Michael Skib commented:

Tested this in a fresh course on staging, and this bug is happening there too.

**Comment by adishy** [comment]
* Seems to be a rather complicated issue that seems to boil down to a problem with views/notification_mailer/grade_released.html.haml

* The problem seems to occur when the NotificationMailer encounters a syntax error with views/notification_mailer/grade_released.html.haml and fails to send an email, causing repeated delayed jobs in the Resque queue "grade_updater", which ends up causing the grade released announcement to be appended multiple times.

* This is supported by the fact that commenting out lines 46 - 49 in performers/grade_update_performer.rb, which deals with sending out the email notification for  a grade release, causes the issue to stop occurring.

* On checking "Failed Jobs" tab in the Resque web console (http://localhost:8282) after multiple notifications appear, the following error message was shown:
```
Worker: aditya-XPS-15-9560:14167 on grade_updater at 25 minutes ago
Class: GradeUpdaterJob 
Arguments: grade_id: 2632
Exception: SyntaxError
Error: /mnt/m/UM_data/gradecraft-development/app/views/notification_mailer/grade_released.html.haml:26: syntax error, unexpected keyword_ensure, expecting end-of-input ensure ^~~~~~
```

* The following is a simplified overview of how a "Grade released" announcement is created:
```
  /assignments/:id/grades/mass_update
  > Services::CreatesManyGroupGrades (Services, LightService::Organizer)
     > IteratesAssignmentGroupsToCreateGrades (Services > Actions, LightService::Action)
         > Services::CreatesGroupGrades (Services, LightService::Organizer)
            > VerifiesGroup (Services > Actions, LightService::Action)
            > IteratesCreatesGrade (Services > Actions, LightService::Action)
               > Services::CreatesGrade (Services, LightService::Organizer)
                  > RunsGradeUpdaterJob (Services > Actions, LightService::Action)
                     > GradeUpdaterJob (ResqueJob::Base)
                        > GradeUpdatePerformer (ResqueJob::Performer)
                             > notify_grade_released
                                > GradeAnnouncement.create @grade
                                    > create_announcement: Announcement.create params  
                                >** NotificationMailer.grade_released(@grade.id).deliver_now 
     > AssertResultFromManyOutcomes
```

**Comment by adishy** [comment]
* Changing the content in views/notification_mailer/grades_released.html.haml to arbotrary placeholder content also causes the issue to stop occurring, (for testing).

* Full stack trace
```
{
  "failed_at": "2019/05/21 14:20:47 EDT",
  "payload": {
    "class": "GradeUpdaterJob",
    "args": [
      {
        "grade_id": 2632
      }
    ]
  },
  "exception": "SyntaxError",
  "error": "/mnt/m/UM_data/gradecraft-development/app/views/notification_mailer/grade_released.html.haml:26: syntax error, unexpected keyword_ensure, expecting end-of-input\n ensure\n ^~~~~~",
  "backtrace": [
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:309:in `module_eval'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:309:in `compile'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:259:in `block (2 levels) in compile!'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:170:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:350:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:258:in `block in compile!'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:246:in `synchronize'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:246:in `compile!'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:158:in `block in render'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `block in instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:354:in `instrument_render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/template.rb:157:in `render'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/rack-mini-profiler-1.0.0/lib/mini_profiler/profiling_methods.rb:78:in `block in profile_method'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/template_renderer.rb:54:in `block (2 levels) in render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/abstract_renderer.rb:44:in `block in instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `block in instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/abstract_renderer.rb:43:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/template_renderer.rb:53:in `block in render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/template_renderer.rb:61:in `render_with_layout'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/template_renderer.rb:52:in `render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/template_renderer.rb:16:in `render'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/renderer.rb:44:in `render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/renderer/renderer.rb:25:in `render'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/rendering.rb:103:in `_render_template'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/rendering.rb:84:in `render_to_body'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/rendering.rb:25:in `render'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/base.rb:919:in `block in collect_responses'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/collector.rb:28:in `custom'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/collector.rb:11:in `html'",
    "/mnt/m/UM_data/gradecraft-development/app/mailers/notification_mailer.rb:112:in `block in send_student_email'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/base.rb:920:in `collect_responses'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/base.rb:828:in `mail'",
    "/mnt/m/UM_data/gradecraft-development/app/mailers/notification_mailer.rb:110:in `send_student_email'",
    "/mnt/m/UM_data/gradecraft-development/app/mailers/notification_mailer.rb:41:in `grade_released'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/base.rb:194:in `process_action'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/callbacks.rb:42:in `block in process_action'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/callbacks.rb:98:in `run_callbacks'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/callbacks.rb:41:in `process_action'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2.1/lib/abstract_controller/base.rb:134:in `process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/rescuable.rb:25:in `block in process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/rescuable.rb:17:in `handle_exceptions'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/rescuable.rb:24:in `process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionview-5.2.2.1/lib/action_view/rendering.rb:32:in `process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/base.rb:607:in `block in process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `block in instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2.1/lib/active_support/notifications.rb:168:in `instrument'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/base.rb:606:in `process'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/message_delivery.rb:123:in `block in processed_mailer'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/message_delivery.rb:122:in `tap'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/message_delivery.rb:122:in `processed_mailer'",
    "/home/aditya/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/actionmailer-5.2.2.1/lib/action_mailer/message_delivery.rb:113:in `deliver_now'",
    "/mnt/m/UM_data/gradecraft-development/app/performers/grade_update_performer.rb:47:in `notify_grade_released'",
    "/mnt/m/UM_data/gradecraft-development/app/performers/grade_update_performer.rb:22:in `block in require_notify_released_success'",
    "/mnt/m/UM_data/gradecraft-development/lib/resque_job/performer.rb:71:in `require_success'",
    "/mnt/m/UM_data/gradecraft-development/app/performers/grade_update_performer.rb:22:in `require_notify_released_success'",
    "/mnt/m/UM_data/gradecraft-development/app/performers/grade_update_performer.rb:9:in `do_the_work'",
    "/mnt/m/UM_data/gradecraft-development/lib/resque_job/base.rb:27:in `perform'"
  ],
  "worker": "aditya-XPS-15-9560:14167:*",
  "queue": "grade_updater"
}
```

**Comment by ebarroso7** [comment]
for some reason, this issue was not closed when tagged in the PR #4284, so i'm closing it now
